src/d/actor/d_a_pw.cpp

```diff
@@ -107,8 +111,253 @@ void first_mode_change(pw_class*) {
 }
 
 /* 00002A90-00003B08       .text action_dousa__FP8pw_class */
-void action_dousa(pw_class*) {
+void action_dousa(pw_class* i_this) {
     /* Nonmatching */
+    daPy_py_c* link = daPy_getPlayerActorClass();
+    float randomValue;
+    switch (i_this->m368) {
+    case 0:
+        anm_init(i_this, 0x23, 7.0, '\x02', 1.0, -1);
+        i_this->m346 = 0;
+        i_this->m368 += 1;
+        break;
+
+    case 0x1:
+        f32 actor_distance = fopAcM_searchActorDistance(i_this, link);
+        if (actor_distance < 500.0f) {
+            anm_init(i_this, 0x10, 3.0, '\0', 1.0, -1);
+            s16 actorYAngle = fopAcM_searchActorAngleY(i_this, link);
+            i_this->m38C = actorYAngle;
+            i_this->current.angle.y = i_this->m38C;
+            i_this->shape_angle.y = i_this->m38C;
+            i_this->m3A4 = -80.0f;
+            i_this->m340 = 1;
+            i_this->m368 += 1;
+        }
+        break;
+
+    case 0x2:
+        i_this->shape_angle.y += 0x1000;
+        cLib_addCalc0(&i_this->m3A4, 1.0, 1.0);
+        i_this->m39A += 7;
+        if (i_this->m39A > 100) {
+            i_this->m39A = 100;
+        }
+        BOOL iVar6 = i_this->m2B4->mFrameCtrl.checkPass(25.0f);
+        if (iVar6 != FALSE) {
+            i_this->m3A4 = 0.0f;
+            first_mode_change(i_this);
+            s16 actorAngleY = fopAcM_searchActorAngleY(i_this, link);
+            i_this->m38C = actorAngleY;
+            JPABaseEmitter* particle =
+                dComIfGp_particle_set(0x82EE, &i_this->m2CC, &i_this->shape_angle);
+            if (particle != NULL) {
+                // JPASetRMtxSTVecfromMtx(i_this->m2B4->mpModel->mpNodeMtxm,
+                // particle->mGlobalRotation, &particle->mGlobalTranslation);
+            }
+            i_this->attention_info.flags = 0;
+            i_this->m368 = 0;
+        }
+        break;
+    case 0x6:
+        i_this->m39A += 10;
+        if (i_this->m39A > 100) {
+            i_this->m39A = 100;
+        }
+        BOOL tempVar10 = TRUE;
+        if (!(i_this->m2B4->mFrameCtrl.getState() & 1) &&
+            i_this->m2B4->mFrameCtrl.getRate() != 0.0f)
+        {
+            tempVar10 = FALSE;
+        }
+        if (tempVar10) {
```

Comment:
```suggestion
        if (i_this->m2B4->isStop()) {
```
can delete the above 6 lines